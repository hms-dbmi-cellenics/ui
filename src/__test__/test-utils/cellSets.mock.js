const mockCellSets = {
  initialLoadPending: false,
  error: false,
  loading: false,
  accessible: true,
  hierarchy: [
    {
      key: 'louvain',
      children: [{ key: 'cluster-a' }, { key: 'cluster-b' }, { key: 'cluster-c' }],
      cellIds: new Set(),
    },
    {
      key: 'sample',
      children: [{ key: 'sample-1' }, { key: 'sample-2' }],
      cellIds: new Set(),
    },
    {
      key: 'scratchpad',
      children: [{ key: 'scratchpad-a' }],
      cellIds: new Set(),
    },
  ],
  properties: {
    'cluster-a': {
      name: 'cluster a',
      key: 'cluster-a',
      cellIds: new Set([0, 1, 6]),
      color: '#01FFFF',
    },
    'cluster-b': {
      name: 'cluster b',
      key: 'cluster-b',
      cellIds: new Set([2, 3]),
      color: '#23FFFF',
    },
    'cluster-c': {
      name: 'cluster c',
      key: 'cluster-c',
      cellIds: new Set([4, 5]),
      color: '#45FFFF',
    },
    'sample-1': {
      name: 'Sample 1',
      key: 'sample-1',
      cellIds: new Set([0, 1, 2]),
      color: '#012FFF',
    },
    'sample-2': {
      name: 'Sample 2',
      key: 'sample-2',
      cellIds: new Set([3, 4, 5]),
      color: '#345FFF',
    },
    'scratchpad-a': {
      cellIds: new Set(['5']),
      key: 'scratchpad-a',
      name: 'New Cluster',
      color: '#5FFFFF',
    },
    louvain: {
      cellIds: new Set(),
      name: 'Louvain clusters',
      key: 'louvain',
      type: 'cellSets',
      rootNode: true,
    },
    sample: {
      cellIds: new Set([6, 7, 8]),
      name: 'Samples',
      key: 'sample',
      type: 'metadataCategorical',
      rootNode: true,
    },
    scratchpad: {
      cellIds: new Set(),
      name: 'Scratchpad',
      key: 'scratchpad',
      type: 'cellSets',
      rootNode: true,
    },
  },
};

export {
  // eslint-disable-next-line import/prefer-default-export
  mockCellSets,
};
