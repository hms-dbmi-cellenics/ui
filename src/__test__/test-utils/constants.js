export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
