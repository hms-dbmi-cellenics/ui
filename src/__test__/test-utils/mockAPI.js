import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
