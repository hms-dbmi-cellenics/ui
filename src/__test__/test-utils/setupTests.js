import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
  Spatial: () => 'Mocked Spatial',
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
