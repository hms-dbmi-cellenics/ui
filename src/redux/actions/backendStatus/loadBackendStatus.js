import handleError from 'utils/http/handleError';
import fetchAPI from 'utils/http/fetchAPI';
import {
  BACKEND_STATUS_LOADING,
  BACKEND_STATUS_LOADED,
  BACKEND_STATUS_ERROR,
} from 'redux/actionTypes/backendStatus';
import endUserMessages from 'utils/endUserMessages';

const loadBackendStatus = (experimentId) => async (dispatch) => {
  dispatch({
    type: BACKEND_STATUS_LOADING,
    payload: {
      experimentId,
    },
  });

  const url = `/v1/experiments/${experimentId}/backendStatus`;
  try {
    const status = await fetchAPI(url);
    dispatch({
      type: BACKEND_STATUS_LOADED,
      payload: {
        experimentId,
        status,
      },
    });

    return status;
  } catch (e) {
    const errorMessage = handleError(e, endUserMessages.ERROR_FETCHING_BACKEND_STATUS);

    dispatch({
      type: BACKEND_STATUS_ERROR,
      payload: {
        experimentId,
        error: errorMessage,
      },
    });
  }
};

export default loadBackendStatus;
