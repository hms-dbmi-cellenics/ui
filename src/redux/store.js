/* eslint-disable global-require */
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { createWrapper } from 'next-redux-wrapper';
import { enableMapSet } from 'immer';

import rootReducer from 'redux/reducers/index';

enableMapSet();

const bindMiddleware = (middleware) => {
  const { composeWithDevTools } = require('redux-devtools-extension');

  // eslint-disable-next-line import/no-extraneous-dependencies
  const { createLogger } = require('redux-logger');

  // do not log server-side redux actions
  middleware.push(createLogger({
    predicate: () => typeof window !== 'undefined',
  }));

  return composeWithDevTools(applyMiddleware(...middleware));
};

const makeStore = () => {
  const store = createStore(
    rootReducer,
    bindMiddleware([thunk]),
  );

  // IF REDUCERS WERE CHANGED, RELOAD WITH INITIAL STATE
  if (module.hot) {
    module.hot.accept('./reducers', () => {
      const createNextReducer = require('./reducers').default;
      store.replaceReducer(createNextReducer(rootReducer));
    });
  }
  return store;
};

const wrapper = createWrapper(makeStore, { debug: false });
export { wrapper, makeStore };
