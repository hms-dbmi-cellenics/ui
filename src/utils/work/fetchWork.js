import { Environment, isBrowser } from 'utils/deploymentInfo';

import cache from 'utils/cache';
import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
import downloadFromS3 from 'utils/work/downloadFromS3';
import waitForWorkRequest from 'utils/work/waitForWorkRequest';

// retrieveData will try to get the data for the given experimentId and ETag from
// the fastest source possible. It will try to get the data in order from:
// 1. Browser cache
// 2. S3 (cache) via signedURL
// 3. Worker
//   3.1. Via socket (small data)
//   3.2. Via S3 (large data) via signedURL
const retrieveData = async (experimentId,
  ETag,
  signedUrl,
  request,
  timeout,
  body,
  dispatch,
  useCache) => {
  // 1. Check if we have the ETag in the browser cache (no worker)
  const cachedData = useCache ? await cache.get(ETag) : null;
  if (cachedData) {
    return cachedData;
  }

  // 2. Check if data is cached in S3 so we can download from the signed URL (no worker)
  if (signedUrl) {
    return await downloadFromS3(body.name, signedUrl);
  }

  // 3. If we don't have signedURL, wait for the worker to send us the data via
  // - the data via socket
  // - the signedURL to download the data from S3
  const { signedUrl: workerSignedUrl, data } = await waitForWorkRequest(
    ETag,
    experimentId,
    request,
    timeout,
    dispatch,
  );

  // 3.1. The worker send the data via socket because it's small enough
  if (data) {
    return data;
  }
  // 3.2. The worker send a signedUrl to download the data
  return await downloadFromS3(body.name, workerSignedUrl);
};

const fetchWork = async (
  experimentId,
  body,
  getState,
  dispatch,
  optionals = {},
) => {
  const {
    extras = undefined,
    timeout = 180,
    broadcast = false,
    onETagGenerated = () => { },
  } = optionals;

  if (!isBrowser) {
    throw new Error('Disabling network interaction on server');
  }

  const { environment } = getState().networkResources;
  if (environment === Environment.DEVELOPMENT && !localStorage.getItem('disableCache')) {
    localStorage.setItem('disableCache', 'true');
  }

  // 1. Contact the API to get ETag and possible S3 signedURL
  const { ETag, signedUrl, request } = await dispatchWorkRequest(
    experimentId,
    body,
    timeout,
    {
      broadcast,
      ...extras,
    },
    dispatch,
  );

  onETagGenerated(ETag);

  const useCache = body.name !== 'GeneExpression';

  // 2. Try to get the data from the fastest source possible
  const data = await retrieveData(
    experimentId, ETag, signedUrl, request, timeout, body, dispatch, useCache,
  );

  // 3. Cache the data in the browser
  if (useCache) {
    await cache.set(ETag, data);
  }

  return data;
};

export default fetchWork;
