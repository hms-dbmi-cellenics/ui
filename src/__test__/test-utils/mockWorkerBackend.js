import * as socketConnectionMocks from 'utils/socketConnection';
import * as fetchWorkResponseMocks from 'utils/work/fetchWork';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/fetchWork', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/fetchWork');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    fetchWork: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
fetchWorkResponseMocks.fetchWork.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
