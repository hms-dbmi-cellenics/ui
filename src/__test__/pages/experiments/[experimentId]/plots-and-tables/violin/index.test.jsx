import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import _ from 'lodash';
import ViolinIndex from 'pages/experiments/[experimentId]/plots-and-tables/violin/index';
import React from 'react';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { loadBackendStatus } from 'redux/actions/backendStatus';
import { makeStore } from 'redux/store';
import fetchWork from 'utils/work/fetchWork';

import markerGenes1 from '__test__/data/marker_genes_1.json';
import paginatedGeneExpressionData from '__test__/data/paginated_gene_expression.json';

import fake from '__test__/test-utils/constants';
import mockAPI, {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
} from '__test__/test-utils/mockAPI';
import createTestComponentFactory from '__test__/test-utils/testComponentFactory';

jest.mock('components/header/UserButton', () => () => <></>);
jest.mock('react-resize-detector', () => (props) => {
  // eslint-disable-next-line react/prop-types
  const { children } = props;
  return children({ width: 800, height: 800 });
});

jest.mock('utils/work/fetchWork');

const mockWorkerResponses = {
  ListGenes: paginatedGeneExpressionData,
  GeneExpression: markerGenes1,
};

const experimentId = fake.EXPERIMENT_ID;
const plotUuid = 'ViolinMain-0';
const multiViewUuid = 'multiView-violin';

const customAPIResponses = {
  [`/plots/${plotUuid}$`]: (req) => {
    if (req.method === 'PUT') return promiseResponse(JSON.stringify('OK'));
    return statusResponse(404, 'Not Found');
  },
};

const defaultResponses = _.merge(
  generateDefaultMockAPIResponses(experimentId),
  customAPIResponses,
);

const defaultProps = { experimentId };

const violinPageFactory = createTestComponentFactory(ViolinIndex, defaultProps);

const renderViolinPage = async (store) => {
  await act(async () => {
    render(
      <Provider store={store}>
        <div width={800} height={800}>
          {violinPageFactory()}
        </div>
      </Provider>,
    );
  });
};

enableFetchMocks();

let storeState = null;

describe('ViolinIndex', () => {
  beforeEach(async () => {
    jest.clearAllMocks();

    fetchWork
      .mockReset()
      .mockImplementation((_experimentId, body) => mockWorkerResponses[body.name]);

    fetchMock.resetMocks();
    fetchMock.mockIf(/.*/, mockAPI(defaultResponses));

    storeState = makeStore();

    // Set up state for backend status
    await storeState.dispatch(loadBackendStatus(experimentId));
  });

  it('Loads controls and elements', async () => {
    await renderViolinPage(storeState);
    expect(screen.getByText(/Gene selection/i)).toBeInTheDocument();

    expect(screen.getByText(/View multiple plots/i)).toBeInTheDocument();
    expect(screen.getByText(/Select data/i)).toBeInTheDocument();
    expect(screen.getByText(/Data transformation/i)).toBeInTheDocument();
    expect(screen.getByText(/Main schema/i)).toBeInTheDocument();
    expect(screen.getByText(/Axes and margins/i)).toBeInTheDocument();
    expect(screen.getByText(/Markers/i)).toBeInTheDocument();
    expect(screen.getByText(/Legend/i)).toBeInTheDocument();
  });

  it('Loads the plot', async () => {
    await renderViolinPage(storeState);

    expect(screen.getByRole('graphics-document', { name: 'Violin plot' })).toBeInTheDocument();
  });

  it('Changes to raw expression', async () => {
    await renderViolinPage(storeState);

    userEvent.click(screen.getByText(/Data transformation/i));

    userEvent.click(screen.getByText('Raw values'));

    expect(screen.getByRole('graphics-document', { name: 'Violin plot' })).toBeInTheDocument();
    expect(storeState.getState().componentConfig[plotUuid].config.normalised).toBe('raw');
  });

  it('Adds a new plot to multi view', async () => {
    await renderViolinPage(storeState);

    userEvent.click(screen.getByText(/View multiple plots/i));

    const searchBox = screen.getByRole('combobox', { name: 'SearchBar' });

    userEvent.type(searchBox, 'cc');

    const option = screen.getByTitle('Ccl5');

    await act(async () => {
      // the element has pointer-events set to 'none', skip check
      // based on https://stackoverflow.com/questions/61080116
      userEvent.click(option, undefined, { skipPointerEventsCheck: true });
    });

    userEvent.click(screen.getByText('Add'));

    await waitFor(() => expect(screen.getAllByRole('graphics-document', { name: 'Violin plot' }).length).toBe(2));

    // check the multi view was expanded
    const multiViewConfig = storeState.getState().componentConfig[multiViewUuid].config;
    expect(_.isEqual([multiViewConfig.nrows, multiViewConfig.ncols], [2, 2])).toBe(true);

    // New plot's config is correct
    expect(storeState.getState().componentConfig['ViolinMain-1']).toMatchSnapshot('new-config');
  });

  it('Adds new plot using same config as previous plot', async () => {
    await renderViolinPage(storeState);

    // Change to raw expression
    userEvent.click(screen.getByText(/Data transformation/i));
    userEvent.click(screen.getByText('Raw values'));

    userEvent.click(screen.getByText(/View multiple plots/i));

    const searchBox = screen.getByRole('combobox', { name: 'SearchBar' });

    userEvent.type(searchBox, 'cc');

    const option = screen.getByTitle('Ccl5');

    await act(async () => {
      // the element has pointer-events set to 'none', skip check
      // based on https://stackoverflow.com/questions/61080116
      userEvent.click(option, undefined, { skipPointerEventsCheck: true });
    });

    userEvent.click(screen.getByText('Add'));

    await waitFor(() => expect(screen.getAllByRole('graphics-document', { name: 'Violin plot' }).length).toBe(2));

    // New plot's config contains normalised set to raw too
    expect(storeState.getState().componentConfig['ViolinMain-0'].config.normalised).toEqual('raw');
  });
  it('Changes the shown gene', async () => {
    await renderViolinPage(storeState);

    userEvent.click(screen.getByText(/Gene selection/i));

    const searchBox = screen.getAllByRole('combobox', { name: 'SearchBar' })[0];

    userEvent.clear(searchBox);
    userEvent.type(searchBox, 'cc');

    const option = screen.getByTitle('Ccl5');

    await act(async () => {
      // the element has pointer-events set to 'none', skip check
      // based on https://stackoverflow.com/questions/61080116
      userEvent.click(option, undefined, { skipPointerEventsCheck: true });
    });

    userEvent.click(screen.getByText('Search'));

    expect(searchBox.textContent).toBe('');

    await waitFor(() => expect(screen.getByRole('graphics-document', { name: 'Violin plot' })).toBeInTheDocument());

    expect(storeState.getState().componentConfig[plotUuid].config.shownGene).toBe('Ccl5');
  });
});
