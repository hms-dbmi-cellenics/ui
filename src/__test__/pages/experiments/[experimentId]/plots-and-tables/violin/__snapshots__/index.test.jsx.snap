// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViolinIndex Adds a new plot to multi view: new-config 1`] = `
{
  "config": {
    "axes": {
      "domainWidth": 1,
      "enabled": true,
      "gridOpacity": 0,
      "gridWidth": 10,
      "labelFontSize": 12,
      "offset": 10,
      "titleFontSize": 13,
      "xAxisRotateLabels": true,
      "xAxisText": "",
      "yAxisText": "",
    },
    "axesRanges": {
      "yAxisAuto": true,
      "yMax": 10,
      "yMin": 0,
    },
    "colour": {
      "gradient": "default",
      "invert": "standard",
      "masterColour": "#000000",
      "reverseColourBar": false,
      "toggleInvert": "#FFFFFF",
    },
    "dimensions": {
      "height": 400,
      "width": 550,
    },
    "fontStyle": {
      "colour": "#000000",
      "font": "sans-serif",
    },
    "kdeBandwidth": 0.3,
    "keepValuesOnReset": [
      "shownGene",
      "title",
    ],
    "labels": {
      "enabled": true,
      "size": 18,
    },
    "legend": {
      "colour": "#000000",
      "enabled": false,
      "position": "right",
      "showAlert": false,
    },
    "marker": {
      "opacity": 5,
      "outline": true,
      "shape": "circle",
      "showOpacity": true,
      "size": 5,
    },
    "normalised": "zScore",
    "selectedCellSet": "louvain",
    "selectedPoints": "All",
    "selectedPointsVisible": true,
    "shownGene": "Ccl5",
    "spec": "1.0.0",
    "statisticsVisible": false,
    "title": {
      "anchor": "start",
      "dx": 0,
      "fontSize": 20,
      "text": "Ccl5",
    },
  },
  "error": false,
  "experimentId": "testae48e318dab9a1bd0bexperiment",
  "loading": false,
  "outstandingChanges": true,
  "plotData": [],
  "plotType": "violin",
}
`;
